### PR TITLE
gmp: modernize more for conan v2

### PR DESCRIPTION
**Pull request by SpaceIm**
Specify library name and version:  **lib/1.0**

---

- [ ] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [ ] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Review by StellaSmith — DISMISSED** [review]

File: recipes/gmp/all/conanfile.py [hunk 1]
```diff
@@ -82,12 +83,11 @@ def build_requirements(self):
             if not self.conf.get("tools.microsoft.bash:path", check_type=str):
                 self.tool_requires("msys2/cci.latest")
         if is_msvc(self):
-            self.tool_requires("yasm/1.3.0")      # Needed for determining 32-bit word size
-            self.tool_requires("automake/1.16.5") # Needed for lib-wrapper
+            self.tool_requires("yasm/1.3.0")
+            self.tool_requires("automake/1.16.5")
```
Comment (prince-chrismc) [inline recipes/gmp/all/conanfile.py:87]: Please put the comments back.

In fact we do not have enough comments and it's desirable some more

File: recipes/gmp/all/conanfile.py [hunk 2]
```diff
@@ -109,11 +109,10 @@ def generate(self):
                 "lt_cv_sys_global_symbol_pipe=cat",  # added to get further in shared MSVC build, but it gets stuck later
             ])
             tc.extra_cxxflags.append("-EHsc")
-            if (self.settings.compiler == "msvc" and Version(self.settings.compiler.version) >= "180") or \
-               (self.settings.compiler == "Visual Studio" and Version(self.settings.compiler.version) >= "12"):
+            if check_min_vs(self, "180", raise_invalid=False):
                 tc.extra_cflags.append("-FS")
                 tc.extra_cxxflags.append("-FS")
-        env = tc.environment() # Environment must be captured *after* setting extra_cflags, etc. to pick up changes
+        env = tc.environment()
```
Comment (prince-chrismc) [inline recipes/gmp/all/conanfile.py:115]: Keep the comments pleases 🙏

Comment (SpaceIm) [reply recipes/gmp/all/conanfile.py:115]: @prince-chrismc recipe is not the appropriate place for this kind of comment. If it's unclear, conan doc should be updated, or package templates if you want. Do you want to update hundreds of CCI recipes to explain how conan functions work?

File: recipes/gmp/all/conanfile.py (hunk 1 above)
Comment (SpaceIm) [reply recipes/gmp/all/conanfile.py:87]: I can put back yasm eventually, but automake comment is useless since it's a common pattern.

**Review by jwillikers — DISMISSED** [review]

**Review by jwillikers — DISMISSED** [review]

**Review by uilianries — APPROVED** [review]

File: recipes/gmp/all/conanfile.py [hunk 3]
```diff
@@ -82,12 +83,11 @@ def build_requirements(self):
             if not self.conf.get("tools.microsoft.bash:path", check_type=str):
                 self.tool_requires("msys2/cci.latest")
         if is_msvc(self):
-            self.tool_requires("yasm/1.3.0")      # Needed for determining 32-bit word size
-            self.tool_requires("automake/1.16.5") # Needed for lib-wrapper
+            self.tool_requires("yasm/1.3.0") # Needed for determining 32-bit word size
+            self.tool_requires("automake/1.16.5")
```
Comment (prince-chrismc) [inline recipes/gmp/all/conanfile.py:87]:
```suggestion
            self.tool_requires("automake/1.16.5") # Needed for lib-wrapper
```

File: recipes/gmp/all/conanfile.py (hunk 2 above)
Comment (prince-chrismc) [reply recipes/gmp/all/conanfile.py:115]:
```suggestion
        env = tc.environment() # Environment must be captured *after* setting extra_cflags, etc. to pick up changes
```

**Review by prince-chrismc — COMMENTED** [review]
We have decided keeping comments are beneficial for reviews since it removes a cognitive load
In addition it has the added benefit for more novice user who are trying to contribute for the first time who may be less familiar.

Our belief: If one person found it useful, then other will as well.

Much to my surprise people use CCI recipes as references when learning Conan so these types of comments have a lot of merit.

**Review by prince-chrismc — APPROVED** [review]

**Comment by SpaceIm** [comment]
I won't argue more because I prefer to see gmp v2 support quickly merged instead of fighting against this decision.

Comments are not beneficial when they refer to widely used or trivial logic, but just noise. Everything is already explained in templates, and in these templates there is https://github.com/conan-io/conan-center-index/blob/951f76d55e7548f12849bf81a43206ed8253a771/docs/package_templates/autotools_package/all/conanfile.py#L16 for a good reason.

**Review by jwillikers — APPROVED** [review]

**Pull request merged**